### PR TITLE
Introduce VectorizedArrayBase and VectorizedArrayIterator

### LINKED ISSUE #8896: Implement range-based iterators for VectorizedArray
This topic came up in PR #8893. The aim is that we can write 
```cpp
for(auto & i : vectorized_array) i++; 
```
as it is familiar from std containers

assign: @peterrum and @kronbichler 

ping: @masterleinad @tamiko

### PR DESCRIPTION
**Pull request by peterrum**
This is a first draft. ~~I am not completely satisfied especially due to the virtual functions, but hopefully someone has a better idea!~~

Closes #8896.

Ping @kronbichler @masterleinad

### CODE REVIEW

**Comment by masterleinad** [comment]
> This is a first draft. I am not completely satisfied especially due to the virtual functions, but hopefully someone has a better idea!

I would use CRTP here.

**Comment by peterrum** [comment]
@masterleinad Thanks! But I solved it differently. Once we have converged, I'll add the log entry and the comments!

File: tests/base/vectorization_iterator_01.output [hunk 1]
```diff
@@ -0,0 +1 @@
+
```
Comment (masterleinad) [inline tests/base/vectorization_iterator_01.output]: Just print "OK".

**Comment by masterleinad** [comment]
No, you solved it using CTRP. :wink:

File: tests/base/vectorization_iterator_01.output (hunk 1 above)
Comment (peterrum) [reply tests/base/vectorization_iterator_01.output]: Done.

**Comment by masterleinad** [comment]
I pushed a version myself. If you find something nicer, I am also happy with that.

**Comment by peterrum** [comment]
@masterleinad Thanks!

@masterleinad @kronbichler I have added some comments and the missing change-log files!

File: doc/news/changes/minor/20191012PeterMunchDanielArndt [hunk 2]
```diff
@@ -0,0 +1,4 @@
+New: Introduce a base class for the VectorizedArray specializations, containing 
+common functionalities (e.g., size, iterator).
```
Comment (kronbichler) [inline doc/news/changes/minor/20191012PeterMunchDanielArndt]: Maybe one suggestion to keep the changelog a bit compact: Can you merge the two changelogs you have and merge it with mine from this week here https://github.com/dealii/dealii/blob/master/doc/news/changes/minor/20191008MartinKronbichler and say that we added some STL-like functionality like an iterator and size?

Comment (peterrum) [reply doc/news/changes/minor/20191012PeterMunchDanielArndt]: Done!

**Pull request merged**